amm/contracts/stable_pool/lib.rs

```diff
@@ -305,47 +293,19 @@ pub mod stable_pool {
             self.pool.tokens[token_id].into()
         }
 
-        /// Update cached rates if expired.
-        ///
-        /// NOTE:
-        /// If the pool contains a token with rate oracle, this function makes
-        /// a cross-contract call to the `RateProvider` contract if the cached rate is expired.
-        /// This means that the gas cost of the contract methods that use this function may vary,
-        /// depending on the state of the expiration timestamp of the cached rate.
-        fn update_rates(&mut self) {
-            let current_time = self.env().block_timestamp();
-            let mut rate_changed = false;
-            for rate in self.pool.token_rates.iter_mut() {
-                rate_changed |= rate.update_rate(current_time);
-            }
-            if rate_changed {
-                Self::env().emit_event(RatesUpdated {
-                    rates: self.pool.token_rates.clone(),
-                });
-            }
-        }
-
-        /// Get updated rates
-        fn get_updated_rates(&self) -> Vec<TokenRate> {
-            let current_time = self.env().block_timestamp();
-            let mut rates = self.pool.token_rates.clone();
-            rates.iter_mut().for_each(|rate| {
-                rate.update_rate(current_time);
-            });
-            rates
-        }
-
         /// Scaled rates are rates multiplied by precision. They are assumed to fit in u128.
         /// If TOKEN_TARGET_DECIMALS is 18 and RATE_DECIMALS is 12, then rates not exceeding ~340282366 should fit.
         /// That's because if precision <= 10^18 and rate <= 10^12 * 340282366, then rate * precision < 2^128.
-        fn get_scaled_rates(&self, rates: &[TokenRate]) -> Result<Vec<u128>, MathError> {
-            rates
-                .iter()
+        fn get_scaled_rates(&mut self) -> Result<Vec<u128>, MathError> {
+            let current_block_no = self.env().block_number();
+            self.pool
+                .token_rates
+                .iter_mut()
                 .zip(self.pool.precisions.iter())
                 .map(|(rate, &precision)| {
-                    rate.get_rate()
+                    rate.get_rate(current_block_no)
                         .checked_mul(precision)
-                        .ok_or(MathError::MulOverflow(114))
+                        .ok_or(MathError::MulOverflow(104))
```

Comment:
Why ID changed? Are you sure that value `104` doesn't repeat in many places? Do you keep the property that error IDs form a sequence `0,1,2,...,K`?

Reply:
It does not repeat. The error IDs 100 - 255 are reserved for errors inside the `stable_pool` contract module